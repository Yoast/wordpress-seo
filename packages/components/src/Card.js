import React, { Fragment } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

/* Yoast dependencies */
import { colors } from "@yoast/style-guide";
import { makeOutboundLink } from "@yoast/helpers";

/* Internal dependencies */
import Banner from "./CardBanner";

const Container = styled.div`
	position: relative;
	display: flex;
	flex-direction: column;
	background-color: ${ colors.$color_white };
	width: 100%;
	box-shadow: 0 2px 4px 0 rgba(0,0,0,0.2);
`;

const HeaderImage = styled.img`
	width: 100%;
	vertical-align: bottom;
`;

const Content = styled.div`
	padding: 12px 16px;
	display: flex;
	flex-direction: column;
	flex-grow: 1;
`;

const HeaderLink = styled.a`
	text-decoration: none;
	color: ${ colors.$color_pink_dark };
	/* IE11 bug header image height see https://github.com/philipwalton/flexbugs#flexbug-5 */
	overflow: hidden;

	&:hover,
	&:focus,
	&:active {
		text-decoration: underline;
		color: ${ colors.$color_pink_dark };
	}

	&:focus,
	&:active {
		box-shadow: none;
	}
`;

const HeaderTitle = styled.h2`
	margin: 16px 16px 0 16px;
	font-weight: 400;
	font-size: 1.5em;
	line-height: 1.2;
	color: currentColor;
`;

const OutboundHeaderLink = makeOutboundLink( HeaderLink );

/**
 * Card component.
 */
class Card extends React.Component {
	/**
	 * Returns the header image and title, with a link if provided.
	 *
	 * @returns {React.Component} The header image.
	 */
	getHeader() {
		if ( ! this.props.header ) {
			return null;
		}

		if ( this.props.header.link ) {
			return (
				<OutboundHeaderLink href={ this.props.header.link }>
					<HeaderImage src={ this.props.header.image } alt="" />
					<HeaderTitle>{ this.props.header.title }</HeaderTitle>
				</OutboundHeaderLink>
			);
		}

		return (
			<Fragment>
				<HeaderImage src={ this.props.header.image } alt="" />;
				<HeaderTitle>{ this.props.header.title }</HeaderTitle>
			</Fragment>
		);
	}

	/**
	 * Gets the banner if a banner text is provided.
	 *
	 * @returns {React.Component} The banner or null.
	 */
	getBanner() {
		if ( ! this.props.banner ) {
			return null;
		}

		return (
			<Banner { ...this.props.banner }>
				{ this.props.banner.text }
			</Banner>
		);
	}

	/**
	 * Renders the component.
	 *
	 * @returns {ReactElement} The rendered component.
	 */
	render() {
		return (
			<Container className={ this.props.className } id={ this.props.id }>
				{ this.getHeader() }
				{ this.getBanner() }
				<Content>
					{ this.props.children }
				</Content>
			</Container>
		);
	}
}

export default Card;

export const FullHeightCard = styled( Card )`
	height: 100%;
`;

Card.propTypes = {
	className: PropTypes.string,
	id: PropTypes.string,
	header: PropTypes.shape( {
		title: PropTypes.string,
		image: PropTypes.string.isRequired,
		link: PropTypes.string,
	} ),
	banner: PropTypes.shape( {
		text: PropTypes.string.isRequired,
		textColor: PropTypes.string,
		backgroundColor: PropTypes.string,
	} ),
	children: PropTypes.any,
};

Card.defaultProps = {
	className: "",
	id: "",
	header: null,
	banner: null,
	children: null,
};
