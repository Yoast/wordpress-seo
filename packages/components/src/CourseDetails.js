/* External dependencies */
import React, { Fragment } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

/* Yoast dependencies */
import { colors } from "@yoast/style-guide";
import { makeOutboundLink, getDirectionalStyle } from "@yoast/helpers";

const CardRegularButton = styled.a`
	color: ${ colors.$color_black };
	white-space: nowrap;
	display: block;
	border-radius: 4px;
	background-color: ${ colors.$color_grey_cta };
	padding: 12px 16px;
	box-shadow: inset 0 -4px 0 rgba(0, 0, 0, 0.2);
	border: none;
	text-decoration: none;
	font-weight: bold;
	font-size: inherit;
	margin-bottom: 8px;

	&:hover,
	&:focus,
	&:active {
		color: ${ colors.$color_black };
		background-color: ${ colors.$color_grey_hover };
	}

	&:active {
		background-color: ${ colors.$color_grey_hover };
		transform: translateY( 1px );
		box-shadow: none;
		filter: none;
	}
`;

const CardUpsellButton = styled.a`
	cursor: pointer;
	color: ${ colors.$color_black };
	white-space: nowrap;
	display: block;
	border-radius: 4px;
	background-color: ${ colors.$color_button_upsell };
	padding: 12px 16px;
	box-shadow: inset 0 -4px 0 rgba(0, 0, 0, 0.2);
	border: none;
	text-decoration: none;
	font-weight: bold;
	font-size: inherit;
	margin-top: 0;
	margin-bottom: 8px;

	&:hover,
	&:focus,
	&:active {
		color: ${ colors.$color_black };
		background: ${ colors.$color_button_upsell_hover };
	}

	&:active {
		background-color: ${ colors.$color_button_hover_upsell };
		transform: translateY( 1px );
		box-shadow: none;
		filter: none;
	}
`;

const CardInfoLink = styled.a`
	font-weight: bold;
`;

const OutboundInfoLink = makeOutboundLink( CardInfoLink );

const ActionBlock = styled.div`
	text-align: center;
`;

const CourseFeatureList = styled.div`
	ul {
		list-style-type: none;
		margin: 0;
		padding: 0;
	}

	li {
		position: relative;
		${ getDirectionalStyle( "margin-left", "margin-right" ) }: 16px;

		&:before {
			content: "âœ“";
			color: ${ colors.$color_green };
			position: absolute;
			font-weight: bold;
			display: inline-block;
			${ getDirectionalStyle( "left", "right" ) }: -16px;
		}
	}
`;

const Details = styled.div`
	margin-bottom: 12px;
	border-bottom: 1px ${ colors.$color_grey } solid;
	flex-grow: 1;
`;

/**
 * CourseDetails component.
 */
class CourseDetails extends React.Component {
	/**
	 * Returns the correct Action Block based on whether an item is a bundle or a single course.
	 *
	 * @param {string} buttonType The type of the button. Either regular or sale.
	 * @param {string} isBundle   Whether the item is a bundle. Default: empty string.
	 *
	 * @returns {ReactElement} The ActionBlock component.
	 */
	getActionBlock( buttonType, isBundle ) {
		const OutboundLinkButton = makeOutboundLink( buttonType );

		// Bundles don't have an OutboundInfoLink and use a different property from the feed for the OutboundLinkButton.
		if ( isBundle === "true" ) {
			return <ActionBlock>
				<OutboundLinkButton href={ this.props.courseUrl }>
					{ this.props.ctaButtonData.ctaButtonCopy }
				</OutboundLinkButton>
			</ActionBlock>;
		}
		return <ActionBlock>
			<OutboundLinkButton href={ this.props.ctaButtonData.ctaButtonUrl }>
				{ this.props.ctaButtonData.ctaButtonCopy }
			</OutboundLinkButton>
			<OutboundInfoLink href={ this.props.courseUrl }>
				{ this.props.readMoreLinkText }
			</OutboundInfoLink>
		</ActionBlock>;
	}

	/**
	 * Renders the component.
	 *
	 * @returns {ReactElement} The rendered component.
	 */
	render() {
		const buttonType = this.props.ctaButtonData.ctaButtonType === "regular" ? CardRegularButton : CardUpsellButton;

		return (
			<Fragment>
				<Details>
					<CourseFeatureList
						dangerouslySetInnerHTML={ { __html: this.props.description } }
					/>
				</Details>
				{ this.getActionBlock( buttonType, this.props.isBundle ) }
			</Fragment>
		);
	}
}

export default CourseDetails;

CourseDetails.propTypes = {
	description: PropTypes.string,
	courseUrl: PropTypes.string,
	ctaButtonData: PropTypes.object,
	readMoreLinkText: PropTypes.string,
	isBundle: PropTypes.string,
};

CourseDetails.defaultProps = {
	description: "",
	courseUrl: "",
	ctaButtonData: {},
	readMoreLinkText: "",
	isBundle: "",
};
