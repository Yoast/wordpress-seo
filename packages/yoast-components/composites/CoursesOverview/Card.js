export { Card as default } from "@yoast/components";
