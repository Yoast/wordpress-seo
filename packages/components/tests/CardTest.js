import React from "react";
import renderer from "react-test-renderer";
import Card, { FullHeightCard } from "../src/Card";
import CourseDetails from "../src/CourseDetails";

test( "The empty Card component matches the snapshot", () => {
	const component = renderer.create(
		<Card />
	);

	const tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );

test( "The FullHeightCard is rendered including some props", () => {
	const component = renderer.create(
		<FullHeightCard className={ "CourseCard_Empty" } id={ "Test01" } />
	);

	const tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );

test( "The FullHeightCard is rendered including the CardBanner and CardDetails components", () => {
	const course = {
		title: "Keyword Research",
		image: "../wp-content/plugins/wordpress-seo/images/sample_course_card_header.png",
		description: "Do you know the essential first step of good SEO? It’s keyword research. In this training, \n" +
		             "you’ll learn how to research and select the keywords that will guide searchers to your pages.",
		courseUrl: "https://yoast.com/academy/keyword-research-training/",
		ctaButtonData: {
			ctaButtonType: "regular",
			ctaButtonCopy: "Start your free trail",
			ctaButtonUrl: "https://yoast.com/cart/?add-to-cart=1311259",
		},
		readMoreLinkText: "Read more about this training »",
		banner: { text: "Free" },

	};
	const header = {
		image: course.image,
		title: course.title,
		link: course.courseUrl,
	};

	const component = renderer.create(
		<FullHeightCard
			className={ "CourseCard" }
			id={ "Test02" }
			header={ header }
			banner={ course.banner }
		>
			<CourseDetails
				title={ course.title }
				description={ course.description }
				courseUrl={ course.courseUrl }
				ctaButtonData={ course.ctaButtonData }
				readMoreLinkText={ course.readMoreLinkText }
			/>
		</FullHeightCard>
	);

	const tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );
