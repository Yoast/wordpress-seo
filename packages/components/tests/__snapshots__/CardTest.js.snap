// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The FullHeightCard is rendered including some props 1`] = `
.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background-color: #fff;
  width: 100%;
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.2);
}

.c2 {
  padding: 12px 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c1 {
  height: 100%;
}

<div
  className="c0 c1 CourseCard_Empty"
  id="Test01"
>
  <div
    className="c2"
  />
</div>
`;

exports[`The FullHeightCard is rendered including the CardBanner and CardDetails components 1`] = `
.c5 {
  border: 0;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute !important;
  width: 1px;
  word-wrap: normal !important;
  -webkit-transform: translateY(1em);
  -ms-transform: translateY(1em);
  transform: translateY(1em);
}

.c6 {
  position: absolute;
  top: 8px;
  left: -8px;
  font-weight: 500;
  color: #fff;
  line-height: 16px;
  background-color: #a4286a;
  padding: 8px 16px;
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.2);
}

.c7 {
  position: absolute;
  top: 40px;
  left: -8px;
  border-top: 8px solid #6c2548;
  border-left: 8px solid transparent;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background-color: #fff;
  width: 100%;
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.2);
}

.c3 {
  width: 100%;
  vertical-align: bottom;
}

.c8 {
  padding: 12px 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c2 {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #a4286a;
  overflow: hidden;
}

.c2:hover,
.c2:focus,
.c2:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: #a4286a;
}

.c2:focus,
.c2:active {
  box-shadow: none;
}

.c4 {
  margin: 16px 16px 0 16px;
  font-weight: 400;
  font-size: 1.5em;
  line-height: 1.2;
  color: currentColor;
}

.c1 {
  height: 100%;
}

.c12 {
  color: #000;
  white-space: nowrap;
  display: block;
  border-radius: 4px;
  background-color: #ddd;
  padding: 12px 16px;
  box-shadow: inset 0 -4px 0 rgba(0,0,0,0.2);
  border: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: bold;
  font-size: inherit;
  margin-bottom: 8px;
}

.c12:hover,
.c12:focus,
.c12:active {
  color: #000;
  background-color: #cecece;
}

.c12:active {
  background-color: #cecece;
  -webkit-transform: translateY( 1px );
  -ms-transform: translateY( 1px );
  transform: translateY( 1px );
  box-shadow: none;
  -webkit-filter: none;
  filter: none;
}

.c13 {
  font-weight: bold;
}

.c11 {
  text-align: center;
}

.c10 ul {
  list-style-type: none;
  margin: 0;
  padding: 0;
}

.c10 li {
  position: relative;
  margin-left: 16px;
}

.c10 li:before {
  content: "✓";
  color: #77b227;
  position: absolute;
  font-weight: bold;
  display: inline-block;
  left: -16px;
}

.c9 {
  margin-bottom: 12px;
  border-bottom: 1px #ddd solid;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

<div
  className="c0 c1 CourseCard"
  id="Test02"
>
  <a
    className="c2"
    href="https://yoast.com/academy/keyword-research-training/"
    rel={null}
    target="_blank"
  >
    <img
      alt=""
      className="c3"
      src="../wp-content/plugins/wordpress-seo/images/sample_course_card_header.png"
    />
    <h2
      className="c4"
    >
      Keyword Research
    </h2>
    <span
      className="c5"
    >
      (Opens in a new browser tab)
    </span>
  </a>
  <span
    className="c6"
  >
    Free
  </span>
  <span
    className="c7"
  />
  <div
    className="c8"
  >
    <div
      className="c9"
    >
      <div
        className="c10"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Do you know the essential first step of good SEO? It’s keyword research. In this training, 
you’ll learn how to research and select the keywords that will guide searchers to your pages.",
          }
        }
      />
    </div>
    <div
      className="c11"
    >
      <a
        className="c12"
        href="https://yoast.com/cart/?add-to-cart=1311259"
        rel={null}
        target="_blank"
      >
        Start your free trail
        <span
          className="c5"
        >
          (Opens in a new browser tab)
        </span>
      </a>
      <a
        className="c13"
        href="https://yoast.com/academy/keyword-research-training/"
        rel={null}
        target="_blank"
      >
        Read more about this training »
        <span
          className="c5"
        >
          (Opens in a new browser tab)
        </span>
      </a>
    </div>
  </div>
</div>
`;

exports[`The empty Card component matches the snapshot 1`] = `
.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background-color: #fff;
  width: 100%;
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.2);
}

.c1 {
  padding: 12px 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

<div
  className="c0"
  id=""
>
  <div
    className="c1"
  />
</div>
`;
